### ISSUE TITLE
9 to 10 frames dropped every 56 seconds

### ISSUE THREAD

**Issue by satysin** [lhc70000/iina#1926]
**System and IINA version:**

- macOS 10.13.6 (17G2208)
- IINA 1.0.0-beta3 Build 86 mpv 0.28.2
- 2018 15" MacBook Pro 2.9Ghz i9, 32GB RAM, 1TB SSD, Intel UHD 620, Radeon Pro 560X

**Expected behavior:**

Video should play back without dropping frames. 

**Actual behavior:**

Every 56 seconds either 9 or 10 frames are dropped. Consistently.

- Crash report:
```

```

- MPV log:
```

```

**Steps to reproduce:**

Always seems to happen no matter what video I play.

- [x] MPV does not have this problem. At least I can't reproduce it in the same way. 

**How often does this happen?**
Every video

**Comment by gitguys** [comment]
Happening here as well.  Lots of skipping playing video.

macOS 10.12.6
IINA 0.0.15.1 Build 69 mpv 0.28.2
2011 17" MacBook Pro i7, 16GB RAM, 1TB SSD, Intel HD Graphics 3000 AMD, Radeon HD 6750M

**Comment by satysin** [comment]
Should also add that VLC does not drop a single frame playing the same video. Both IINA and VLC are using the Intel integrated graphics which should be more than enough to handle video playback. I get dropped frames with standard H.264 720p videos when the machine is running no other user apps.

**Comment by lhc70000** [comment]
Which version of mpv did you tested with?

I have no idea why this problem happened, but it should be either a mpv bug or related to IINA's rendering. We'll upgrade libmpv and update the rendering mechanism a little bit in the next release (beta 4), and if you don't mind, please also report if this issue persists in the next version.

**Comment by satysin** [comment]
mpv 0.28.2 installed via Homebrew so same mpv version as included with IINA from what I can tell. This is why I thought it is an IINA issue rather than an mpv one (but I could be wrong 😬)

The strange thing is that it drops frames at exactly 56 seconds and always 9 or 10 frames. 

As a small update I did manage to watch a 20 minute video with no dropped frames earlier today so I am not sure if it is some kind of cache issue that shows itself after a certain amount of time? 

What can I do to help you find the problem? What log files or anything else would help?

**Comment by satysin** [comment]
Small update. I just watched the same video file (42 minute TV show) in IINA (version details as above) and a configless mpv (0.28.2, installed via Homebrew) and IINA had lots of dropped frames (again in groups of around 10) whereas mpv had none.

**Comment by inflation** [comment]
It’s a known issue if you’re using any graphic-audio sync options. You can enable the logging in advanced settings and post here. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by satysin** [comment]
Not configured any graphic-audio sync options. Just stock IINA config. What options specifically should I look at? Have enabled logging, will watch a video then post the log file here.

**Comment by satysin** [comment]
[iina.log](https://github.com/lhc70000/iina/files/2312245/iina.log)
[mpv.log](https://github.com/lhc70000/iina/files/2312246/mpv.log)

Log files as requested.

I _think_ I might have worked out when it happens... It seems if I switch from the iGPU to the dGPU the issue appears. Obviously this is done by macOS as needed but I was able to repro it by toggling 'Automatic graphics switching' in Energy Saver off and on. When on the dGPU (auto switching unchecked) there are no dropped frames. It also seems that when on the iGPU all is fine _until_ the dGPU is activated. So perhaps something to do with that?

Let me know if the log files are helpful at all. I paused then closed IINA a few seconds after the frames dropped so the end of the log file should be very close to where the drop happens.

**Comment by inflation** [comment]
Yep, that’s the issue: #1542. We will wait and see if the new rendering mechanism works (probably would😅).

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by satysin** [comment]
@inflation still an issue in beta 4 :(

**Comment by JayBrown** [comment]
Yep, still getting the issue as well (newest IINA version). If you load a video fresh, start playback, the bug does not occur. If you pause the video, switch to other applications & processes, and leave IINA idle for some time (10 minutes? 15?), and then return to IINA to continue playback, the video playback gets stuck every 1 minute or so for a couple of frames. (Audio is fine.) Doesn't seem to happen if you leave IINA idle only for a short period of time.

It might be that the dGPU also has something to do with it. Will test that as well.

**Comment by JayBrown** [comment]
Tested: left IINA, launched Photoshop which toggled dGPU for a couple of times, then went back to IINA, and the video dropouts started to happen, not immediately, but after 2 or 3 minutes.

**Comment by lhc70000** [comment]
I missed something in #1542. Will upload a new build here when implemented it.

**Comment by satysin** [comment]
Any news of an ETA with the fix?

**Comment by wrz0819** [comment]
还是没有任何解决方法吗？beta4貌似也没有很好地解决这个问题

**Comment by inflation** [comment]
@wrz0819 Should be fixed in `develop` branch

**Comment by an4rchy** [comment]
Having the same issue on the latest version. Video freezes/drops every 29-35secs (eyeballed the timestamp).

**Comment by lhc70000** [comment]
I managed to reproduce this issue for a few times after switched to integrated GPU by unplugging external monitor, but most of the time I failed to trigger it. I'll try to find other ways that can reproduce stably when have more time.

**Comment by satysin** [comment]
Here are two ways that reliably repro the issues

1. Uncheck and check the automatic graphics switching option in System Preferences\Energy Saver

2. Using Google Chrome go to maps.google.com as this will cause Chrome to switch to the Radeon GPU. This was how I originally found the issue as I was looking up a location in the film I was watching 😬

**Comment by lhc70000** [comment]
Thanks for the clue, now I'm able to reproduce it by unchecking and then checking the automatic graphics switching option (frame drops start to happen when switched to integrated GPU), regardless of whether hardware decoding is on.

After some logging I found that the rendering process, i.e. `mpv_render_context_render` got stuck for a few calls (spent >100ms to render a frame), and seems that **mpv has the same problem**.

Have you all tried to reproduce this issue with mpv? If so, we should report it to them.

**Comment by satysin** [comment]
Hmm I tried to repro this with mpv originally but it didn't seem to happen. This is why I raised the bug here and not with mpv. 

However I just tried and you're right. It seems to be a bug in mpv not IINA :(

**Comment by satysin** [comment]
Correct @Akemi once the GPU switches the dropped frames continue even if the system continues to use the new GPU for the rest of playback without switching GPU again. 

It seems once the GPU switches _something_ gets out of whack and the dropped frames continue until IINA is terminated and relaunched. 

For the moment I am working around this by disabling automatic graphics switching before I load IINA so macOS only uses the dGPU and therefore cannot switch between GPUs. It works fine with the exception of the dGPU eating the battery.

**Comment by JayBrown** [comment]
You can restrict graphics to the iGPU if you use [**gfxCardStatus**](https://github.com/steveschow/gfxCardStatus/releases), the fork by @steveschow. Version 2.4.4i doesn't seem to work properly on Mojave on my hardware (i9 MBP), but v2.4.3i does, and when I'm on the road, I use the iGPU exclusively. (Note: this fork of gfxCardStatus isn't being developed further, and I don't know if anyone else is currently working on another version.)

**Comment by satysin** [comment]
@JayBrown Thanks for the heads up about 2.4.3i working. I tried 2.4.4i and, like you said, it didn't work so I didn't look into it anymore.

It is a shame it is abandoned as it was a helpful little program. Although I feel Apple should have an option to manually select the GPU rather than just "disable automatic graphics switching" and just using the dGPU.

**Comment by AlexanderOMara** [comment]
Was this ever reported upstream? I seems to be experiencing this myself.

**Comment by satysin** [comment]
@AlexanderOMara Not by me sorry, I lack the expertise. Perhaps one of the IINA developers did though?

**Comment by phyllisstein** [comment]
I can still reproduce this bug in v1.0.7-beta1, build 115, with mpv 0.32.0, mostly when playing large H264-encoded videos. To echo @satysin’s comment [above](https://github.com/iina/iina/issues/1926#issuecomment-449728502), disabling automatic graphics switching reliably stops IINA from skipping.

**Comment by JayBrown** [comment]
By the way: check out **[gSwitch](https://github.com/CodySchrank/gSwitch)** by @CodySchrank … seems to work fine on my 2018 MBP. What would be great, until the mpv bug is fixed, is a command-line version, so we could create an IINA-Launcher that reads & then locks the current GPU state before launching IINA itself.

**Comment by Akemi** [comment]
> ... until the mpv bug is fixed ...

i don't think there is a bug report for this on the mpv issue tracker (so very unlikely anything gets fixed). also it's most likely something inherent from shitty macOS opengl. so it will most likely never be fixed. not as a workaround/fix in mpv neither by apple.

also since the context creation and drawing is completely handled by the client it's also likely that mpv is not at fault here, and it could possibly fixed in iina itself, if at all.

**Comment by JayBrown** [comment]
I've retooled a small NSWorkspace listener, and I'm now using it to listen for IINA, and if IINA launches, it will check if dynamic switching is enabled, and if yes, then it will use the `pmset` command to switch the GPU to integrated-only (when on battery), and to discrete-only (when on AC). It seems to be fast enough for the switch to finish before IINA loads the video.

**Comment by microfx** [comment]
Any news on this? Having problems with a brand new Macbook Pro 16" using IINA with integrated GPU to not get deaf from the fans. Shoooould be possible with this configuration, right?

**Comment by Akemi** [comment]
i am doing the work for mpv right now to utilise moltenvk which utilise metal https://github.com/mpv-player/mpv/pull/7482

**Comment by heartbleeded** [comment]
Any update on this? I still have extreme lagging problem while watching video from IINA 1.1.2 under integrated Intel in fullscreen mode. VLC doesn't have any problem though.

**Comment by TheCrawlingKhaos** [comment]
This happens to me as well. It only happens if I try to play a second video without first closing the player after watching the first video. This only started happening after I upgraded to a M1 Mac. (Currently running Big Sur 11.4 with 8GB of RAM) If I close the player the issue does not occur again until I try to play a second video consecutively.
